File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        int N = nums.length;
        int[] max = new int[N + 1];

        max[N] = 0;
        max[N - 1] = nums[N - 1];

        for (int i = N - 2; i >= 0; --i) {
            max[i] = Math.max(max[i + 1], max[i + 2] + nums[i]);
        }

        return max[0];
    }
}

File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        rob1, rob2 = 0, 0

        # rob1, rob2, nums[i], nums[i + 1], ... , nums[n - 1]
        for num in nums:
            rob1, rob2 = rob2, max(num + rob1, rob2)
        return rob2